docs/proposal/circuit_breaker.md

```diff
@@ -0,0 +1,374 @@
+				log.Errorf("cluster %s %s flush failed: %v", name, cluster.ApiStatus, err)
+			}
+		} else if cluster.GetApiStatus() == core_v2.ApiStatus_DELETE {
+			err := maps_v2.ClusterDelete(name)
+			if err == nil {
+				delete(cache.apiClusterCache, name)
+				delete(cache.resourceHash, name)
+				cache.hashName.Delete(name)
+			} else {
+				log.Errorf("cluster %s delete failed: %v", name, err)
+			}
+		}
+	}
+}
+```
+You can see that we introduce a hashName to map string to integer.
+
+Here we also add a new field `id` to cluster:
+
+```protobuf
+message Cluster {
+  enum LbPolicy {
+    ROUND_ROBIN = 0;
+    LEAST_REQUEST = 1;
+    RANDOM = 3;
+  }
+
+  core.ApiStatus api_status = 128;
+  string name = 1;
+  uint32 id = 2;
+  uint32 connect_timeout = 4;
+  LbPolicy lb_policy = 6;
+
+  endpoint.ClusterLoadAssignment load_assignment = 33;
+  CircuitBreakers circuit_breakers = 10;
+}
+```
+
+To monitor current active tcp connections, we need to create a `BPF_MAP_TYPE_SK_STORAGE` map:
+
+```c
+struct cluster_sock_data {
+    __u32 cluster_id;
+};
+
+struct {
+    __uint(type, BPF_MAP_TYPE_SK_STORAGE);
+    __uint(map_flags, BPF_F_NO_PREALLOC);
+    __type(key, int);
+    __type(value, struct cluster_sock_data);
+} map_of_cluster_sock SEC(".maps");
+```
+
+We can manage the lifecycle of a socket based on it.
+
+Then, we can follow the flow chart below:
+
+<div align="center">
+    <img src="./pics/kmesh_circuit_breaker_flow.png" />
+</div>
+
+We can monitor socket operations in eBPF "sockops" hooks. First, we judge whether the active connections for a cluster have reaches max thresholds. If so, we should reject the connection (how to do so is still TBD). Otherwise, we allow the connection, and handle it based on the type of socket op.
+
++ TCP_DEFER_CONNECT:
+
+    We will enter sockops traffic control flow in this branch. It will trigger a series of chain calls, finally reaching `cluster_manager` (check the following image).
+
+    <div align="center">
+        <img src="./pics/kmesh_ads_mode_sockops_flow.png" width="50%" />
+    </div>
+
+    We will get cluster information here (e.g, cluster id). We can store the cluster id in the `cluster_sock_data`. At this stage, we have bound the cluster to the socket.
+
+    We can do so by calling this function below in `cluster_manager`:
+
+    ```c
+    static inline void on_cluster_sock_bind(struct bpf_sock *sk, const char* cluster_name) {
+        BPF_LOG(DEBUG, KMESH, "record sock bind for cluster %s\n", cluster_name);
+        struct cluster_sock_data *data = NULL;
+        if (!sk) {
+            BPF_LOG(WARN, KMESH, "provided sock is NULL\n");
+            return;
+        }
+
+        data = bpf_sk_storage_get(&map_of_cluster_sock, sk, 0, BPF_LOCAL_STORAGE_GET_F_CREATE);
+        if (!data) {
+            BPF_LOG(ERR, KMESH, "record_cluster_sock call bpf_sk_storage_get failed\n");
+            return;
+        }
+
+        bpf_strncpy(data->cluster_name, BPF_DATA_MAX_LEN, (char *)cluster_name);
```

Comment:
This is introduced in kernel 5.17, but we should support 5.10

Reply:
https://github.com/kmesh-net/kmesh/blob/cce29bcdc8e709d53d8c08d66bbe9875c40b0453/bpf/kmesh/ads/include/tail_call.h#L88-L92

but i found it also use `bpf_strncpy`, maybe we need an alternative

Reply:
IC, this is defined with kernel patch